### ISSUE TITLE
[critical] Performance issue in latest nightly

### ISSUE THREAD

**Issue by gayanper** [spring-projects/sts4#294]
There are lot of elastic threads are created for monitoring local running applications when editing a file.

**Comment by gayanper** [comment]
[elastic.zip](https://github.com/spring-projects/sts4/files/3207070/elastic.zip)

**Comment by kdvolder** [comment]
Looking at the stacktraces, this rings a bell. On Windows calling methods on `HotSpotVirtualMachine` sometimes gets blocked indefinitely. In order to keep things 'alive' we have added a 'timeout' utility wrapping around all these calls. I think this is working and keeps things rolling along but unfortunately it still doesn't terminate the blocking call... (it just unblocks the thread 'consuming' the result of that call).

I think what's happening is that the blocked threads pile up over time and since this is a 'elastic' thread pool new threads are created to be able to execute stuff. (If it weren't for these extra threads, things would grind to complete stop).

I'm not sure there's a lot we can do about this. Maybe I can try to implement a variant of the 'timeout' utility that tries to agrressively call Thread.destroy.  I'll experiment with that bit. But I'll have to first setup a Windows dev environment, to try to reproduce the problem first.

**Comment by kdvolder** [comment]
I've tried to reproduce this for a bit on a Windows 10 VirtualBox VM but haven't had any luck with it.

@gayanper Is this a new issue for you? You mention you seen it in latest nightly, but question is... is it new to the nightly build or was it there before too?  I don't think we've done really any changes lately to how the processes are being scraped for live hover data. So that would be a bit surprising.

**Comment by gayanper** [comment]
@kdvolder well i found this in the latest nightly while i was profiling for slowness in code completion. So i'm not sure whether this was the same before.

**Comment by kdvolder** [comment]
Unfortunately, I don't think there is much that we can really do to avoid this pile-up of stuck threads. The only method that can forcibly terminate threads in JVM is `Thread.stop`, but it is deprecated because it is notorious for causing problems with leaving things in broken state. The only other thing I can think of would be to create a totally separate process for scraping boot app processes so that we can periodically kill this process and recreate it as kind of drastic 'cleanup'. 

Both of these options would be quite a lot of effort to implement. for using thread.stop, I think we basically would have to implement our own ExecutorService more or less from scratch because none of the default ones that I know of make use of Thread.stop to kill off threads (probably for good reason). 

And the separate process setup is also quite complex, and has the potential to actually degrade performance as well because its usually more expensive to create and communicate with an external process (using sockets or whatnot).

So before we really go and try any of that, I'd like to get a better sense of how serious of a problem these extra threads are and in how far they are actually responsible for some of the slowness that you are experiencing. As an experiment, maybe you could try disabling the live boot hint support in your setup and see if that makes a big difference in performance. If yes, then I guess it is some evidence that the threads may have something to do with it.

**Comment by gayanper** [comment]
@kdvolder leaving this much of threads running wild can cause the resource starvation in the development machine isn't it ? It might not be directly effecting but in long run the development machine might be in trouble.

**Comment by kdvolder** [comment]
Yes, but since these threads are basically stuck and there is simply no mechanic in the JVM to kill them, there isn't much that can be done (short of completely terminating the language server process).

**Comment by kdvolder** [comment]
@gayanper I'm not so sure what we should do, as said before there are some ways we could try to cope, but as its quite complex (complete rewrite of how we scrape processes, in a much more complex way than it works now), I'd like to get a sense of how serious the impact is of these threads in your day-to-day use. Do you experience it as a real problem, or is it just something you notice when collecting a stackdump. Admittedly, seeing this many threads in a dump looks bad and will draw attention :-). But the question is if it actually eats up so CPU and memory to the point it becomes noticeable as you work?

**Comment by gayanper** [comment]
@kdvolder I don't think this is a huge problem i think we can close this issue :)

**Comment by kdvolder** [comment]
Okay, closing this then. We can always re-open and reconsider if needed.
